src/evaluator.rs

```diff
@@ -0,0 +1,802 @@
+        Token::Slash => Ok(Rc::new(Object::Int(left / right))),
+        Token::Lt => Ok(Rc::new(Object::Bool(left < right))),
+        Token::Gt => Ok(Rc::new(Object::Bool(left > right))),
+        Token::Eq => Ok(Rc::new(Object::Bool(left == right))),
+        Token::Neq => Ok(Rc::new(Object::Bool(left != right))),
+        _ => Err(EvalError{message: format!("unknown operator {}", operator)}),
+    }
+}
+
+fn eval_block(block: &BlockStatement, env: Rc<RefCell<Environment>>) -> EvalResult {
+    let mut result = Rc::new(Object::Null);
+
+    for stmt in &block.statements {
+        let res = eval_statement(stmt, Rc::clone(&env))?;
+
+        match *res {
+            Object::Return(_) => return Ok(res),
+            _ => result = res,
+        }
+    }
+
+    Ok(result)
+}
+
+fn eval_statement(stmt: &Statement, env: Rc<RefCell<Environment>>) -> EvalResult {
+    match stmt {
+        Statement::Expression(exp) => eval_expression(&exp.expression, env),
+        Statement::Return(ret) => {
+            let value = eval_expression(&ret.value, env)?;
+            Ok(Rc::new(Object::Return(Rc::new(object::Return{value}))))
+        },
+        Statement::Let(stmt) => {
+            let exp = eval_expression(&stmt.value, Rc::clone(&env))?;
+            let obj = Rc::clone(&exp);
+            env.borrow_mut().set(stmt.name.clone(), obj);
+            Ok(exp)
+        }
+    }
+}
+
+fn eval_program(prog: &Program, env: Rc<RefCell<Environment>>) -> EvalResult {
+    let mut result = Rc::new(Object::Null);
+
+    for stmt in &prog.statements {
+        let res = eval_statement(stmt, Rc::clone(&env))?;
+        let v = Rc::clone(&res);
+
+        match &*v {
+            Object::Return(r) => return Ok(Rc::clone(&r.value)),
+            _ => result = res,
+        }
+    }
+
+    Ok(result)
+}
+
+fn eval_prefix_expression(operator: &Token, right: Rc<Object>) -> EvalResult {
+    match *operator {
+        Token::Bang => eval_bang_operator_expression(right),
+        Token::Minus => eval_minus_prefix_operator_expression(right),
+        _ => Err(EvalError{message:format!("unknown prefix operator {}", operator)}),
+    }
+}
+
+fn eval_bang_operator_expression(right: Rc<Object>) -> EvalResult {
+    Ok(Rc::new(match *right {
+        Object::Bool(true) => Object::Bool(false),
+        Object::Bool(false) => Object::Bool(true),
+        Object::Null => Object::Bool(true),
+        _ => Object::Bool(false),
+    }))
+}
+
+fn eval_minus_prefix_operator_expression(right: Rc<Object>) -> EvalResult {
+    match *right {
+        Object::Int(val) => {
+            Ok(Rc::new(Object::Int(-val)))
+        },
+        _ => Err(EvalError{message: format!("unknown operator: -{:?}", right)}),
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use object::Object;
+
+    #[test]
+    fn eval_integer_expression() {
+        struct Test<'a> {
+            input: &'a str,
+            expected: i64,
+        }
+        let tests = vec![
```

Comment:
Is the table test idiom used by Rustaceans or is this frowned upon?

Reply:
Yeah it is common pattern used, you can even look into [quickcheck](/BurntSushi/quickcheck) & [proptest](altsysrq/proptest) which are really popular libraries that helps doing those table tests in some ways.